### ISSUE TITLE
DNS Provider: Godaddy - failed to get certificate

### ISSUE THREAD

**Issue by TheGammelgalopper** [go-acme/lego#519]
Hey guys,
I have run into quite a complex issue...
I use Caddy as a purely internal proxy. However a domain from Godaddy is being used internally to get LE certificates.
I did already run my config (Caddyfile) successfully for about 17 subdomains and the domain itself. Now I wanted to add additional subdomains for services which were added to the network. But running it now via this command "sudo GODADDY_API_KEY=xxxxxxx GODADDY_API_SECRET=xxxxxxxxx caddy"results in the following error:

```
2018/03/22 18:56:36 [media.halnet.cloud] failed to get certificate: Error presenting token: Could not create record [{"type":"TXT","name": "_acme-challenge.media.halnet", "data":"uYcrqEGjcrf6iRdWdRTvT2coO6R7wB5l3n7Mvb9uSlE", "ttl":600}]; Status: 404; Body: {"code":"UNKNOWN_DOMAIN", "message": "The given domain is not registered, or does not have a zonefile", "name":"_Class"}
```

The domain clearly is registered! It already worked for other subdomains after all...

In the Godaddy admin panel all other zonefiles were created successfully in the following format: “_acme-challenge.host”. But no new zonefiles are being created...
Even if I manually enter a new zonefile for “_acme-challenge.media” I can’t start caddy, as it still gives the same error…

running caddy with -log zone.log gives the following log file:

```
2018/03/26 20:45:51 [INFO][media.halnet.cloud] acme: Obtaining bundled SAN certificate
2018/03/26 20:45:51 [INFO][media.halnet.cloud] AuthURL: https://acme-v01.api.letsencrypt.org/acme/authz/j9_EDSFAzwccGXrY6fIR_pckpPzlfQ9zspt8utrYBZc
2018/03/26 20:45:51 [INFO][media.halnet.cloud] acme: Trying to solve DNS-01

```

I really have no clue what's going on! Any Help is appreciated. Especially @smerschjohann could you please look into this? It is not urgent, but I would like to get this to work at some time...

Original post in Caddy forum: https://caddy.community/t/cant-get-certificates-for-certain-domains/3543/3
and in Godaddy forum: https://www.godaddy.com/community/Managing-Domains/Can-not-create-TXT-record-via-caddy-for-certain-subdomains/td-p/99894

**Comment by TheGammelgalopper** [comment]
Ok I did a fresh install of Caddy on a new system. I now wanted to go the wildcard-route...

I changed my Caddyfile to something like this:
```

(standard) {
  gzip
  log /var/log/caddy/access.log
  errors /var/log/caddy/errors.log
  tls {
    dns godaddy
	wildcard
  }
}
halnet.cloud {
  tls {
    dns godaddy
  }
  proxy / 10.10.10.40
}
pdu.halnet.cloud {
  proxy / 10.10.10.2
  import standard
}
jarvis.halnet.cloud {
  proxy / https://10.10.10.3/
  import standard
}
```

However it still throws errors...

Log shows this:
```
sysadmin@caddy:/usr/local/bin$ cat on.log
2018/04/07 18:31:54 [INFO][halnet.cloud] acme: Obtaining bundled SAN certificate
2018/04/07 18:31:55 [INFO][halnet.cloud] AuthURL: https://acme-v02.api.letsencrypt.org/acme/authz/6cHNf31eKuCZdCcRbJ8Iyae1k2lfz5GD3q673Jx8ZUw
2018/04/07 18:31:55 [INFO][halnet.cloud] acme: Could not find solver for: http-01
2018/04/07 18:31:55 [INFO][halnet.cloud] acme: Trying to solve DNS-01

```

The error itself:
`Activating privacy features... 2018/04/07 18:31:56 [halnet.cloud] failed to get certificate: Error presenting token: Could not create record [{"type":"TXT","name":"_acme-challenge.halnet","data":"hH_4TDDO_ML4w1MVSSI6zAzP_CPrSLmiiAb-6xTEZ50","ttl":600}]; Status: 404; Body: {"code":"UNKNOWN_DOMAIN","message":"The given domain is not registered, or does not have a zone file","name":"_Class"}`

I have two ideas of what might be wrong:

1. My network can't solve that DNS-01 and/or http-01. How would I test that?
2. acme or lego doesn't know what to do with the .cloud top-level-domain....

Can somebody pls help me here? I really don't know what is going on.... And I have the feeling that this is not just a stupid IBM error....

**Comment by mholt** [comment]
First, stop -- before you run into LE's rate limits. [Use their staging endpoint](https://caddyserver.com/docs/automatic-https#testing) before doing any more testing. :wink:

**Comment by TheGammelgalopper** [comment]
Well I alraedy ran into their limits once... (https://caddy.community/t/use-my-own-domain-purely-internally/3460/18) Not a huge deal... The erros back then looked diferrent ;)

Additionally I shouldn't run into them! With the wildcard directive I am only asking for two certs: one for the domain and one wildcard for all the subdomains...

**Comment by TheGammelgalopper** [comment]
So there are a couple of things i did....

1. Disabled all custom internal dns... I had something set up so that all requests to halnet.cloud are being routed to my caddy box. I turned that off for the sake of getting the certs... At some point I will have to turn it back on without braking the setup...

2. For testing I used the "-ca https://acme-staging-v02.api.letsencrypt.org/directory" thing...

3. Requested a new Godaddy production api key

4. I was now getting a new error: `2018/04/08 15:47:40 [halnet.cloud] failed to get certificate: acme: Error 403 - urn:ietf:params:acme:error:unauthorized - Incorrect TXT record "nELHLWqF__redacted__c8J0" found at _acme-challenge.halnet.cloud`... So I went into the DNS control panel at Godaddy and manually copied that string from above in the value field of the txt record... That field originally just showed "null"...

5. Then caddy threw this error: `2018/04/08 15:56:26 open /var/log/caddy/access.log: no such file or directory`... So I simply created that directory manually

And suddenly caddy starts successfully... In the logs I see that the wildcard cert was saved successfully!

**However!** There still is a problem at hand!

I then tried doing the same thing without the staging environment. Again I had to copy over the new value of the txt record into the dns control panel at godaddy...

Should I create a new issue for this? This seems like a bug to me...

Another question: Assuming that I won't change this setup for the next time, when will I have to copy over the txt record value? I guess not each time I start caddy. But each time I make changes to the Caddyfile?
